File: leetcode/src/main/java/com/thanos/algorithm/ruby/longest_sub_string_length.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    m = Hash.new
    bidx, max = 0, 0

    s.each_char.with_index do |char, idx|
        if (m[char] != nil) then
            max = (idx - bidx) > max ? (idx - bidx) : max
            bidx = (m[char] + 1) > bidx ? (m[char] + 1) : bidx
        end
        m[char] = idx
    end
    max = (s.length - bidx) > max ? (s.length - bidx) : max
    max
end

File: leetcode/src/main/golang/longest_sub_string_length.go
func lengthOfLongestSubstring(s string) int {
    m := make(map[byte]int)
    pIdx, idx, max := 0, 0, 0

    for ; idx < len(s); idx++ {
        if v, ok := m[s[idx]]; ok {
            max = getMax(idx - pIdx, max)
            pIdx = getMax(v + 1, pIdx)
        }
        m[s[idx]] = idx
    }
    max = getMax(idx - pIdx, max)
    return max
}

func getMax(a, b int) int {
    if a < b {
        return b
    }
    return a
}